pkg/routing/kvstore/kvstore.go

```diff
@@ -0,0 +1,183 @@
+package kvstore
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/nats-io/nats.go"
+	"github.com/nats-io/nats.go/jetstream"
+	"github.com/pkg/errors"
+	"github.com/samber/lo"
+
+	"github.com/bacalhau-project/bacalhau/pkg/models"
+	"github.com/bacalhau-project/bacalhau/pkg/routing"
+)
+
+type NodeStoreParams struct {
+	BucketName     string
+	ConnectionInfo interface{}
+}
+
+type NodeStore struct {
+	js jetstream.JetStream
+	kv jetstream.KeyValue
+}
+
+func NewNodeStore(params NodeStoreParams) (*NodeStore, error) {
+	url, ok := params.ConnectionInfo.(string)
+	if !ok {
+		return nil, errors.New("invalid connection info provided to KV Node Store")
+	}
+
+	// The connection we get from NATS is thread-safe (see https://pkg.go.dev/github.com/nats-io/nats.go#Conn)
+	// so no need to wrap everything in a mutex
+	nc, err := nats.Connect(url)
+	if err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("failed to connect to nats network at %s", url))
+	}
+
```

Comment:
I created a second connection deliberately to avoid swamping a single connection with too much traffic.  The connections are thread-safe, and so don't need locks on them, but it feel right to try and keep different types of traffic separate.  I can definitely use the main one if you prefer?